torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -759,7 +883,8 @@ inline bool operator==(const BucketKey& lhs, const BucketKey& rhs) {
 std::vector<std::vector<size_t>> compute_bucket_assignment_by_size(
```

Comment:
nit: This function is called in two places: 1) DDP ctor and 2) when rebuilding mapping. Its behaviors in these two situations are different and are governed by the value in `tensor_indices` arg. I wonder if we can split this into two methods to make it clear? E.g, we can do the following:

1. Add a new `default_bucket_assignment_by_size` which does not take a `tensor_indices` vector
2. Modify `compute_bucket_assignment_by_size` to let it always expect `tensor_indices` to contain the real indices
3. `default_bucket_assignment_by_size` calls into `compute_bucket_assignment_by_size` by providing a default vector, where `tensor_indices[i] = i`. As this is only called in ctor, the overhead of creating a new vector should be OK.
4. Expose `default_bucket_assignment_by_size` to Python.

Reply:
this sounds a good idea, but I checked default_bucket_assignment_by_size is also called by broadcast_coalesced(), the broadcast_coalesced() is called in forward() per iteration if broadcast_buffer is enabled. that means every iteration needs to create a tensor_indices vector, it may be costly. 

On the other hand, I feel like keeping tensor_indices to be empty in default in compute_bucket_assignment_by_size() is probably OK, it implicitly means tensor_indices[i] = i, which is default bucket assignment. let me make comments more clear?

Reply:
default_bucket_assignment_by_size is called in broadcast_coalesced as well, broadcast_coalesced is called per iteration, adding tensor_indices every iteration may be costly.

on the other hand, I feel like it is not too bad to make tensor_indices to be empty in default, representing  tensor_indices[i] = i. I'm making comments more clear, please let me know whether it makes sense to you?

Reply:
I see, you mean the buffer sync in Python `_sync_params` is calling into this? In this case, let's leave it as is.